### DISCUSSION TITLE
Implementing StructOps Support

### DISCUSSION THREAD

**Discussion by takehaya** [cilium/ebpf#1502] [Ideas]
StructOps are used in sched_ext, mptcp, and tcp congestion control. It appears that cilium/ebpf does not support them. Why hasn't this been addressed?

If possible, I would like to work on this. Could you guide me on where to start?

**Comment by lmb** [comment]
I'm happy to help, the problem is that I don't really know how struct_ops work. Maybe you can explain them to me?

- Which program type do they use?
- How are they attached?

**Reply by shun159** [reply]
@takehaya @lmb 

I wanted to make sched_ext work with this library, so I wrote a patch like the following. 
https://github.com/shun159/ebpf/tree/struct_ops_draft

For example, if we use `sched_ext`, we need to resolve user structure `sched_ext_ops` and the corresponding kernel structure `bpf_struct_ops_sched_ext_ops`, and then processes them.

**Reply by shun159** [reply]
@takehaya @lmb @ti-mo 

> How are they attached?


The attachment of struct_ops can be through the following steps:

1. define the kernel structure:
   Within the BPF program, define the specific kernel structure (e.g., `sched_ext`) using BTF information.

2. create of the struct_ops map:
   Create a struct_ops map using the map type `BPF_MAP_TYPE_STRUCT_OPS`. This map represents the kernel structure and holds the function implementations provided from user space.

3. set data to the map:
   Load `BPF_PROG_TYPE_STRUCT_OPS` programs and  write the file descriptors (FDs) of the programs corresponding to each member (func pointers) to the appropriate offset positions within the `kern_vdata`. For regular data members, set the necessary byte date as needed.

4. load the map:
   Load the map into the kernel. When the struct_ops map is loaded, the kernel registers the provided structure and function implementations.
    
5. attach the map:
   we can use `AttachRawLink` in this patch. we pass the fd of the map as `programFD`, and then the program will be attached.
   
```go
package main

import (
	"fmt"
	"log"
	"os"
	"os/signal"
	"syscall"

	"github.com/pkg/errors"

	"github.com/cilium/ebpf"
	"github.com/cilium/ebpf/link"
	"github.com/cilium/ebpf/rlimit"
)

//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target amd64 bpf bpf/sched_ext.bpf.c -- -Wno-compare-distinct-pointer-types -Wno-int-conversion -Wnull-character -g -c -O2 -D__KERNEL__

func main() {
	// Allow the current process to lock memory for eBPF resources.
	if err := rlimit.RemoveMemlock(); err != nil {
		log.Fatal(err)
	}

	objs := bpfObjects{}
	if err := loadBpfObjects(&objs, nil); err != nil {
		log.Fatalf("loading objects: %v", errors.WithStack(err))
	}
	defer objs.Close()

	m := objs.SimpleSched1
	link.AttachRawLink(link.RawLinkOptions{
		ProgramFd: m.FD(),
		Target:    0,
		Attach:    ebpf.AttachStructOps,
	})

	stopper := make(chan os.Signal, 1)
	signal.Notify(stopper, os.Interrupt, syscall.SIGTERM)

	<-stopper

	log.Print("quit sched_ext")
}
```

**Reply by takehaya** [reply]
Oh, sorry I missed your comment :bow:

That's great. It would be nice if you could attach it!

**Reply by shun159** [reply]
@takehaya Yeah, I can attach to the following program. But association with .bss and .rodata etc. might be required (I haven't worked on yet). Therefore, currently I can't use bpf_printk or access global variables, so the code look like this.

```c
#include "vmlinux.h"

#include <bpf/bpf_helpers.h>
#include <bpf/bpf_tracing.h>

#include "scx_helper.h"

char _license[] SEC("license") = "GPL";

#define SHARED_DSQ 0

__s32
BPF_STRUCT_OPS(simple_select_cpu, struct task_struct *p, __s32 prev_cpu, __u64 wake_flags)
{
  bool is_idle = false;
  __s32 cpu;

  cpu = scx_bpf_select_cpu_dfl(p, prev_cpu, wake_flags, &is_idle);
  if (is_idle)
    scx_bpf_dispatch(p, SCX_DSQ_LOCAL, SCX_SLICE_DFL, 0);

  return cpu;
}

void
BPF_STRUCT_OPS(simple_dispatch, __s32 cpu, struct task_struct *p)
{
  scx_bpf_consume(SHARED_DSQ);
}

void
BPF_STRUCT_OPS(simple_stopping, struct task_struct *p, bool runnable)
{
  /*
   * Scale the execution time by the inverse of the weight and charge.
   *
   * Note that the default yield implementation yields by setting
   * @p->scx.slice to zero and the following would treat the yielding task
   * as if it has consumed all its slice. If this penalizes yielding tasks
   * too much, determine the execution time by taking explicit timestamps
   * instead of depending on @p->scx.slice.
   */
  p->scx.dsq_vtime += (SCX_SLICE_DFL - p->scx.slice) * 100 / p->scx.weight;
}

__s32
BPF_STRUCT_OPS_SLEEPABLE(simple_init)
{
  return scx_bpf_create_dsq(SHARED_DSQ, -1);
}

SEC(".struct_ops.link")
struct sched_ext_ops simple_sched1 = {.select_cpu = (void *)simple_select_cpu,
                                      .dispatch = (void *)simple_dispatch,
                                      .stopping = (void *)simple_stopping,
                                      .init = (void *)simple_init,
                                      .timeout_ms = 20000,
                                      .flags = 1,
                                      .name = "simple_sched1"};
```

**Reply by shun159** [reply]
> association with .bss and .rodata etc. might be required

@ti-mo 
could you please tell me how can we associate struct_ops program with special maps?

**Reply by ti-mo** [reply]
> how can we associate struct_ops program with special maps?

What do you mean? These are bpf programs like any other, not sure why they couldn't use global data maps. Whether the verifier allows printk in this context is a different story, haven't tried that.

**Reply by shun159** [reply]
@ti-mo

Thank you for your reply!

For example, if there's a reference to `.rodata` within the program, the question is do we have a way to associate an instruction which has `PseudoMapValue` as src and a FD of a Map pointing to rodata. For instance, in the case of a `ProgramSpec` like the following.
```
ps: simple_enable:
	; bpf_printk("hogehoge");
	0: LoadMapValue dst: r1, fd: 0 off: 0 <.rodata> <<<<<<<<<<<<<<<<<<<<<<<<<<
	2: MovImm dst: r2 imm: 9
	3: Call FnTracePrintk
	; BPF_STRUCT_OPS(simple_enable, struct task_struct *p)
	4: Exit
```

**Reply by ti-mo** [reply]
I'm confused, is something not working? This plumbing all happens behind the scenes.

**Reply by shun159** [reply]
Hmm...I see, so that is probably a problem on my end. Thanks!

**Reply by ti-mo** [reply]
Could you please share some error output so we're sure we're talking about the same thing? It's hard to help without any real information.

**Reply by shun159** [reply]
The following is an error message when loading a program that includes references to `.rodata`. 
```
2024/10/01 21:59:42 loading objects: field SimpleSched1: map simple_sched1: failed to load struct_ops program: simple_enable cannot load program without loading its whole collection: instruction 0: unsatisfied map reference
```
However, I may have figured out the cause. It's probably that the program should be loaded from NewCollectionWithOptions(), but I'm not doing that now.

**Reply by ti-mo** [reply]
> It's probably that the program should be loaded from NewCollectionWithOptions(), but I'm not doing that now.

Exactly. :slightly_smiling_face: This is not an error related to StructOps, it goes for all ELFs. See https://ebpf-go.dev/concepts/loader/#loadandassign if you're only interested in a subset of map or program handles.

**Reply by shun159** [reply]
Thank you! 

I will reconsider the implementation a bit while referring to the documentation. (I didn't know there was such a good resource available 😄)

**Reply by shun159** [reply]
Thanks. For now, I implemented it according to the document and your advise, confirmed that it works, and submitted a draft PR. https://github.com/cilium/ebpf/pull/1603

**Comment by shun159** [comment]
@takehaya could you reopen here? I'd like to continue discussing how to proceed.

**Reply by takehaya** [reply]
I believe it would be best to work on this within the draft PR you created, rather than here. 
Since you’re the one working on it, I’d prefer not to receive notifications for every update, if possible.

**Reply by ti-mo** [reply]
@shun159 Given you already wrote the code, we may as well discuss on the PR itself.

**Reply by shun159** [reply]
@ti-mo 
let me explanatory slides. The implementation basically based on the content described in the slides. I hope it will be helpful. As for how to split the PRs, let me discuss while also referring to the draft PR. thanks!

https://speakerdeck.com/shun159/introduction-structops

**Reply by shun159** [reply]
First, we'll set our current goal as enabling linking of `.struct_ops.link`.
The requirements for supporting `StructOps` can be broadly categorized into two:

1. ELF reading process
2. Loading process

For 1, the expected scope of work includes classification of `structOpsSection`, generation of `MapSpec`, and association of `ProgramSpec` and `MapSpec`. Also, we will consider the points you pointed out below. I'm thinking of placing necessary helper functions and `StructOpsSpec` in `internal/struct_ops`.

>MapSpec and btf.Spec are important pieces of exported API and the bar is high for extending them. Avoid using them for storing arbitrary state that isn't useful for the caller.
Likewise, avoid trivial, overly-specific helpers on btf.Spec. The ones added here are basically struct ops business logic. Conder making them free functions in a separate package, e.g. internal/structops.

Question:
I'm wondering how to pass the state processed by `LoadCollectionSpecFromReader()` to LoadAndAssign(). While it might be possible to create some mapping by extending `CollectionSpec`, what do you think about extending `CollectionSpec`? For example, like adding a map to `CollectionSpec` to map `StructOpsSpec` and `MapSpec` that hold things like a list of pointers to `ProgramSpec` and `btfID`, as shown below:

```diff
+type StructOpsSpec struct {
+	ProgramSpecs []*ProgramSpec
+	KernFuncOff  []uint32
+	/* e.g. struct tcp_congestion_ops in bpf_prog's btf format */
+	Data []byte
+	/* e.g. struct bpf_struct_ops_tcp_congestion_ops in
+	 *      btf_vmlinux's format.
+	 * struct bpf_struct_ops_tcp_congestion_ops {
+	 *	[... some other kernel fields ...]
+	 *	struct tcp_congestion_ops data;
+	 * }
+	 * kern_vdata-size == sizeof(struct bpf_struct_ops_tcp_congestion_ops)
+	 * bpf_map__init_kern_struct_ops() will populate the "kern_vdata"
+	 * from "data".
+	 */
+	KernVData []byte
+	TypeId    btf.TypeID
+	Btf       *btf.Spec
+}

// CollectionSpec describes a collection.
type CollectionSpec struct {
	Maps     map[string]*MapSpec
	Programs map[string]*ProgramSpec

	// Variables refer to global variables declared in the ELF. They can be read
	// and modified freely before loading the Collection. Modifying them after
	// loading has no effect on a running eBPF program.
	Variables map[string]*VariableSpec

	// Types holds type information about Maps and Programs.
	// Modifications to Types are currently undefined behaviour.
	Types *btf.Spec

	// ByteOrder specifies whether the ELF was compiled for
	// big-endian or little-endian architectures.
	ByteOrder binary.ByteOrder
	
+	StructOpsSpecs map[string]*StructOpsSpec
}

```

**Reply by shun159** [reply]
for 2, basically after the program and map are loaded, we expect to need processing that generates `kern_vdata` and writes `kern_vdata` as a value to the map when attaching.
Whether to link using `AttachRawLink` like in the draft PR, or prepare a dedicated function like `AttachStructOps` and perform `kern_vdata` writing and Link at that time, can be decided later.
(It might be better to prepare a StructOps-specific Attach function)

**Reply by shun159** [reply]
I plan to split the PR into at least 4-5 parts. We will add helpers in each PR as needed. (Please let me submit doc/ and example/ in the final PR)

@ti-mo 
I'm looking forward to your comment. have a good weekend :smile:

**Reply by ti-mo** [reply]
Hi @shun159, I haven't forgotten, I'll try to allocate some time next week.

**Reply by shun159** [reply]
I'd like to resume this if that sounds good to everyone.
please let me know if there are any questions. thx

**Comment by lmb** [comment]
@shun159 thanks for pinging. Seems like this is going to be quite an involved feature, doing it in small PRs is definitely the way to go.

Some feedback upfront from looking at the draft PR:

- The ELF reader must not depend on the environment. Think of it as a pure function from ELF to *Spec. `createStructOpsMap` violates this by looking at the kernel types. The part that needs kernel types has to happen when creating the map.
- Like Timo said, avoid modifying exported API at all costs. You may be able to stuff the metadata you need into MapSpec.Contents. Something like `MapSpec.Contents[uint32(0)] = structOpsMeta{}`.
- I think this is doable without changing the exported API.

Can we figure out together in which order it would make sense to land things? As far as I understand, getting struct ops into the kernel involves the following:

User:
1. Declare a `struct` type in their .c which includes a bunch of function pointers. The struct may also contain other, non-function pointer data.
2. Load that collection
3. Take the map fd (?) and attach that via RawLink.

Library:
1. Discover struct ops from `__section(".struct_ops.link")` in the ELF. Each variable in that section is a separate map of type StructOpsMap.
2. Associate programs with the map. This seems to work [similarly to how ProgramArrays are populated](https://github.com/cilium/ebpf/blob/main/testdata/btf_map_init.c).
3. Load the struct ops map, which in turn loads necessary programs.
4. Insert programs into the map. The difference to ProgramArray is that program fds are just stuffed into a []byte instead of being inserted via map_update_elem (?). There is some trickiness here because we also need to account for differences between the kernel and the user type (so things might be at distinct offsets, etc.)

If that is correct the following order / PRs would make sense to me.

1. Create a StructOps map from a hand-crafted MapSpec. Assume that the user type == kernel type, so no complicated copying around based on BTF.
2. Support user type != kernel type, by implementing something like `func extendType(sourceData []byte, source, dest Type) ([]byte, error)`.
3. Load a StructOpsMap from a CollectionSpec via collectionLoader.
4. Read StructOps MapSpec from ELF and output into CollectionSpec.

**Reply by shun159** [reply]
@lmb  
Thank you for going through that old, very big patch.  
I really appreciate the detailed comments!

* I’ve summarized below covered by the current draft patch.
  * making a **hand‑written `StructOpsMap`** when the user struct and kernel struct are the same;
  * putting program FDs straight into the map’s value inside the loader;
  * parsing **`.struct_ops` / `.struct_ops.link`** sections and adding them to `CollectionSpec`.
* The only thing still missing is support for the case where **user and kernel layouts are different** (your `extendType()` idea).

I’ve outlined the next steps below.

Because the old patch is far from today’s `main`, I will start again from the current `main` branch and send a series of small PRs:

1. **Add `StructOpsMap` type**  
   * just the enum value and a tiny helper;  
   * no public‑API change (extra data stays in `MapSpec.Contents` for now).

2. **Load a hand‑written `StructOpsMap`** (same layout)  
   * proves the new map type can be loaded and used.

3. **Add `extendType()`**  
   * copies / rearranges data when the layouts are different;  
   * includes unit tests for reordered or padded fields.

4. **Use it in `CollectionLoader`**  
   * loader fills the map and writes FDs into the value blob;  
   * works end‑to‑end with `LoadAndAssign`.

5. **Update the ELF reader**  
   * automatically create `StructOpsMap` from `.struct_ops` / `.struct_ops.link`;  
   * connect programs via relocations.

6. **Clean‑up and docs**  
   * tidy up unsafe code, add examples, update documentation.

I’ll send **PR　step 1** soon so you can check the direction before we move on.
Thanks again for your help!

**Reply by shun159** [reply]
Hi, @lmb 

This is the first one. I believe this covers the following (also step 1 and 2 in the above):

https://github.com/cilium/ebpf/compare/main...shun159:ebpf:feature/struct-ops-map_1

> Create a StructOps map from a hand-crafted MapSpec. Assume that the user type == kernel type, so no complicated copying around based on BTF.

**Reply by shun159** [reply]
@lmb The patch above is quite small, but the next patch will be much larger because it will add the mapping between the user-space structures and the in-kernel structures.
If you’re happy with the above patch, I’d like to submit a PR and move forward with the remaining work.

**Reply by shun159** [reply]
@lmb Would it be easier for you if I open a PR with the patch in its current state, so we can start the review and discussion?

**Reply by lmb** [reply]
Hi @shun159, I didn't see your message, sorry. Yes, if you want us to review something the way to go is to create a PR.

Please be aware that I won't be able to spend much time on this until September.

**Reply by shun159** [reply]
Thanks, @lmb. 
Since the change is quite large, I’ll submit a series of MRs. First, I’ll modify link.go so that it accepts the ProgramFd from a struct_ops map.
https://github.com/cilium/ebpf/compare/cilium:ebpf:main...shun159:ebpf:feature/struct-ops-map_1

**Reply by shun159** [reply]
I've submitted a PR for link so that it accept programFd option from struct_ops map.
https://github.com/cilium/ebpf/pull/1844

**Reply by shun159** [reply]
also, here is the PR so that create a "hand-crafted" StructOpsMap MapSpec with metadata  
https://github.com/cilium/ebpf/pull/1845

**Comment by shun159** [comment]
@takehaya Thank you for your patience, we’ve finally added StructOps support to cilium/ebpf.
Let's give it a try!
